File: flutter/challenge/lib/model/categoryModel.dart
class CategoryModel {
  String id;
  String name;
  int relevance;

  String iconImageUrl;
  String largeImageUrl;
  String mediumImageUrl;
  String smallImageUrl;

  List<CategoryModel> subcategories;

  CategoryModel(
      {this.id,
      this.name,
      this.relevance,
      this.iconImageUrl,
      this.largeImageUrl,
      this.mediumImageUrl,
      this.smallImageUrl,
      this.subcategories});

  factory CategoryModel.fromJson(Map<String, dynamic> json) {
    return CategoryModel(
        id: json["id"],
        name: json["name"],
        relevance: json["relevance"],
        iconImageUrl: json["iconImageUrl"],
        largeImageUrl: json["largeImageUrl"],
        mediumImageUrl: json["mediumImageUrl"],
        smallImageUrl: json["smallImageUrl"],
        subcategories: json["subcategories"] == null
            ? null
            : json["subcategories"]
                .map<CategoryModel>((i) => CategoryModel.fromJson(i))
                .toList());
  }

  String getImage() {
    String s = this.iconImageUrl ??
        this.smallImageUrl ??
        this.mediumImageUrl ??
        this.largeImageUrl;
    return s;
  }
}

File: src/main/java/cl/talavera/api/data/model/CategoryModel.java
package cl.talavera.api.data.model;

import cl.talavera.api.core.domain.Category;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.Data;

import java.util.List;

@Data
@JsonIgnoreProperties(ignoreUnknown = true)
public class CategoryModel {
    private String id;
    private String name;
    private List<CategoryModel> subcategories;
    private int relevance;
    private String iconImageUrl;
    private String largeImageUrl;
    private String mediumImageUrl;
    private String smallImageUrl;


    public static Category CategoryFrom(CategoryModel response) {
        return Category.builder()
                .name(response.getName())
                .relevance(response.getRelevance())
                .small(response.getSmallImageUrl())
                .build();
    }
}
